KAFKA-6144: Allow state stores to serve stale reads during rebalance

vinothchandar opened the pull request:
KIP-535 Implementation

*Summary of testing strategy*
- [x] Unit tests for standby metadata
- [x] Integration test for querying standbys, during rebalance
- [x] Local testing
- [x] Unit tests around APIs used for allLocalOffsetLags() API
- [x] Integration test for lag APIs; LagInfo class
- [x] Final test coverage good; all new methods; even some love for deprecated APIs
- [ ] Streams System tests passing (1 test failing; currently triaging)
- [ ] Special cases like source topic optimization; 
- [ ] Update KIP with latest 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vinothchandar committed:
[KAFKA-6144]: Allow serving interactive queries from in-sync Standbys

     - Rough initial cut, with basic HA working. Many more changes needed still
     - [TODO] Introduce a new version for AssignmentInfo & support version probing
     - [TODO] Implement API changes in KafkaStreams (deprecation, new classes, new methods)
     - [TODO] Tests using metadataForKey(), redone over new APIs. New tests with standbys involved.
     - [TODO] Tests need to be (re)written, including version probing ones

brary committed:
Adding KeyQueryMetadata support for reads during rebalancing

brary committed:
Resolving code review comments

brary committed:
Adding changes related to fetching offset lag on query

brary committed:
Reverting offset lag function to as discussed in KIP

brary committed:
Adding support to fetch checkpointed offsets and calculate lag

brary committed:
Minor checkstyle fixes and returning the planned offsetLag map

vinothchandar committed:
Minor cleanups

 - Reimplemented KafkaStreams#allLocalOffsetLags()
 - Moved KeyQueryMetadata to org.apache.kafka.streams so we can import publicly
 - Minor renames, whitespace fixes

vinothchandar committed:
Re-implement KafkaStreams#allLocalOffsetLags() by fetching endOffsets from Kafka

 - Re-implement the assignment protocol to serialize the topic index just once
 - Code cleanups, small bug fixes

vinothchandar committed:
Fix issue with reading standbyHostPartition maps for version <= 5

 - Also cleanup non-essential code formatting changes

vinothchandar committed:
Adding tests for standby streams metadata

vinothchandar commented:
```
Test Result (1 failure / +1)
org.apache.kafka.streams.integration.OptimizedKTableIntegrationTest.shouldApplyUpdatesToStandbyStore
```

Is the flaky test, that also fails locally. Will see if there is a quick fix. Otherwise, it seems to pass. 

@vvcephei  This does not have the tests for the lag API yet. But if you could make a quick pass and see if we are close, tht would be awesome.. Will iterate and make one final pass based on your comments..

brary committed:
Fixing shouldApplyUpdatesToStandbyStore test case

brary commented:
> ```
> Test Result (1 failure / +1)
> org.apache.kafka.streams.integration.OptimizedKTableIntegrationTest.shouldApplyUpdatesToStandbyStore
> ```
> 
> Is the flaky test, that also fails locally. Will see if there is a quick fix. Otherwise, it seems to pass.
> 
> @vvcephei This does not have the tests for the lag API yet. But if you could make a quick pass and see if we are close, tht would be awesome.. Will iterate and make one final pass based on your comments..

@vinothchandar I have fixed the test. It was failing as it used the assumption that standby stores don't return data so had to change the logic to differentiate between active and replica.

AlanConfluent commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1112,126 @@ public void cleanUp() {
+        final long unknownPosition = 0;
+
+        // Obtain the current positions, of all the active-restoring and standby tasks
+        for (final StreamThread streamThread : this.threads) {
```
Do you need to do some synchronization while going through all of these threads and their metadata?

Within StreamThread and TaskManager, it doesn't seem like some of these fields are threadsafe.  Might make sense to use add locking and/or return an immutable copy of some of these maps.

I'm not too familiar with the code just yet, so please tell me if I'm missing something.

brary replied:
The information which we are fetching from StreamThreads is basically the state of each task i.e. whether an active/standby is in running/restoring/created state. And these state maps are concurrent hashmaps so I think should be fine. Let me know if I am missing any non thread safe use of these threads.

vinothchandar replied:
I made another pass.. but I believe all the maps fetched are concurrent hash maps, which should support such access.

 - streamThread.allStandbyTasks() ultimately calls AssignedTasks#allTasks, that populates a new list from two concurrent maps
- standbyTask.checkpointedOffsets() gets a new unmodifiable map as well, will make couple maps concurrent as well in ProcessorStateManager, for sake of avoiding future issues if some code were to call remove() on them
- streamThread.restoringTaskIds(), streamThread.allStreamsTasks() all grab values from  concurrent maps. 
- activeTask.restoredOffsets() actually does have one corner case, around suspending using ChangelogReader#remove. Will make the corresponding map concurrent as well. 

if you see any other gap or a better way than fishing like this let us know :)

AlanConfluent commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1112,126 @@ public void cleanUp() {
+        final Map<String, Map<Integer, Long>> localOffsetLags = new HashMap<>();
+        final Map<TopicPartition, Long> standbyChangelogPositions = new HashMap<>();
+        final Map<TopicPartition, Long> activeChangelogPositions = new HashMap<>();
+        final long unknownPosition = 0;
```
You could make this a constant rather than a local variable.  Isn't zero technically a valid offset?  Maybe -1?

AlanConfluent commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1112,126 @@ public void cleanUp() {
+                final String storeName = streamsMetadataState.getStoreForChangelogTopic(topicPartition.topic());
+                final long offsetPosition;
+                if (activeChangelogPositions.containsKey(topicPartition)) {
+                    // if unknown, assume it's positioned at the tail of changelog partition
```
For standbys, you conservatively assume that unknowns are at the head, which means that if another standby is known, that will be picked to read from.  I assume that we'll always read from actives, regardless of their lag, is that right?  What's the benefit of making this less conservative assumption here?

vinothchandar replied:
I am not sure we will always read from actives.. For e.g currently you can have an active be in restoring state, while standbys are farther ahead (just that no further processing will happen until active comes back up). KIP-441 is changing this..

AlanConfluent commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1112,126 @@ public void cleanUp() {
+                }
+                localOffsetLags.put(storeName, partitionToOffsetLag);
+            });
+        } catch (final Exception e) {
```
Do you want to move this try catch around `adminClient.listOffsets(offsetSpecMap)` to make this as localized as possible?  You might accidentally catch another exception here (like the IllegalStateException) and have a misleading message.

brary replied:
I have added a localized catch and removed try-catch for the lines it was used earlier as I don't think IllegalStateException should throw a StreamsException. Let me know if my understanding is not correct.

AlanConfluent commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -1110,20 +1126,28 @@ public void onAssignment(final Assignment assignment, final ConsumerGroupMetadat
+            case 6:
+                processVersionTwoAssignment(logPrefix, info, partitions, activeTasks, topicToPartitionInfo, partitionsToTaskId);
+                partitionsByHost = info.partitionsByHost();
+                standbyPartitionsByHost = info.standbyPartitionByHost();
```
A question for my knowledge:  This looks like it sends both the active and standby assignments to participating nodes, which makes sense.  How did nodes know they were assigned to be a standby before this addition?

brary replied:
The new addition that we have is sending the map of standby partitions corresponding to each host. Earlier each machine had the same map for only active partitions. But the assignments for each machine itself i.e. its own active and standby tasks were already there in the AssignmentInfo. They just didn't know which machine is holding which standby task.

AlanConfluent reviewed (commented):
I know this is a WIP, but made a few comments while reading it over.

brary committed:
Addressing code review comments

brary committed:
Fixing the checkstylese

vinothchandar committed:
Changing few other maps to concurrent maps for safer iteration

vinothchandar committed:
Adding tests for KafkaStreams#allLocalOffsetLags()

vinothchandar committed:
Introduce a LagInfo class that holds both raw offset position, as well as offset lag

  - Improved javadocs for new class and KeyQueryMetadata
  - Redid the tests, based on new APIs

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1113,125 @@ public void cleanUp() {
+        for (final StreamThread streamThread : this.threads) {
+            for (final StandbyTask standbyTask : streamThread.allStandbyTasks()) {
+                final Map<TopicPartition, Long> changelogPartitionLimits = standbyTask.checkpointedOffsets();
+                standbyTask.changelogPartitions().forEach(topicPartition ->
```
Just something to bear in mind as you make more changes: it'd be better to just stick to regular for loops instead of using the `forEach` method on collections. On `Stream`s it's more natural, but for regular collections, the for loop is just fine.

vinothchandar replied:
I dont have strong opinions on this. but like to understand the reasoning behind the suggestion? is it because a lot of the code follows this style?

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1113,125 @@ public void cleanUp() {
+        // Obtain the current positions, of all the active-restoring and standby tasks
+        for (final StreamThread streamThread : this.threads) {
+            for (final StandbyTask standbyTask : streamThread.allStandbyTasks()) {
+                final Map<TopicPartition, Long> changelogPartitionLimits = standbyTask.checkpointedOffsets();
```
"limits" is a confusing name for this, since StandbyTask has a `offsetLimits` field also.

vinothchandar replied:
fair. will rename to `checkpointedOffsets`..

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1113,125 @@ public void cleanUp() {
+        log.info("Current changelog positions, for active: " + activeChangelogPositions + " standby:" + standbyChangelogPositions);
+        final Map<TopicPartition, OffsetSpec> offsetSpecMap = new HashMap<>();
+        Stream.concat(activeChangelogPositions.keySet().stream(), standbyChangelogPositions.keySet().stream())
+            .forEach(topicPartition ->  offsetSpecMap.put(topicPartition, OffsetSpec.latest()));
```
Just a style question, why not collect the result as a map?

vinothchandar replied:
We can. will do

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1113,125 @@ public void cleanUp() {
+            for (final StreamTask activeTask : streamThread.allStreamsTasks()) {
+                final boolean isRestoring = restoringTaskIds.contains(activeTask.id());
+                final Map<TopicPartition, Long> restoredOffsets = activeTask.restoredOffsets();
+                activeTask.changelogPartitions().forEach(topicPartition -> {
```
We need to make sure we also have tests with the optimization to re-use source topics as changelogs, to make sure all this stuff works under that case as well.

vinothchandar replied:
Okay. I heard of the optimization, regardless of how I feel about it, I agree we should cover this. Can you point me at some tests that do cover it.. I can seed off that that..

vvcephei replied:
I don't know off the top of my head what they are, although if you have an integration test, it seems pretty straightforward to run it both with and without optimizations. The javadoc on `table()` explains the optimization: `org.apache.kafka.streams.StreamsBuilder#table(java.lang.String, org.apache.kafka.streams.kstream.Consumed<K,V>, org.apache.kafka.streams.kstream.Materialized<K,V,org.apache.kafka.streams.state.KeyValueStore<org.apache.kafka.common.utils.Bytes,byte[]>>)`

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -0,0 +1,102 @@
+     * Sentinel to indicate that the KeyQueryMetadata is currently unavailable. This can occur during rebalance
+     * operations.
+     */
+    public final static KeyQueryMetadata NOT_AVAILABLE = new KeyQueryMetadata(new HostInfo("unavailable", -1),
```
should be `public static final` (mis-sorted)

vinothchandar replied:
duh. how come checkstyle did not catch this.. anyways, will change

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KeyQueryMetadata.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return set of standby {@link HostInfo} or a empty set, if no standbys are configured
+     */
+    public Set<HostInfo> getStandbyHosts() {
```
method is unused, indicating a missing test.

There are unused methods in other classes as well (eg LagInfo). Can you do a pass at some point and make sure everything gets used or removed?

vinothchandar replied:
yes. I will do this before calling the final review.. I am still working on a test using real lag values, just figured out how to stop a streams thread at a certain state transition using the KafkaStreamsWrapper.. that test will exercise  the LagInfo methods..

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -302,10 +302,10 @@
     @SuppressWarnings("WeakerAccess")
     public static final String APPLICATION_SERVER_CONFIG = "application.server";
-    private static final String APPLICATION_SERVER_DOC = "A host:port pair pointing to an embedded user defined endpoint that can be used for discovering the locations of state stores within a single KafkaStreams application";
+    private static final String APPLICATION_SERVER_DOC = "A host:port pair pointing to an embedded user defined endpoint that can be used for state store discovery and interactive queries within a single KafkaStreams application";
```
Thanks for fixing these docs

```suggestion
    private static final String APPLICATION_SERVER_DOC = "A host:port pair pointing to a user-defined endpoint that can be used for state store discovery and interactive queries on this KafkaStreams instance.";
```

vinothchandar replied:
Will incorporate

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -71,6 +73,15 @@ public String toString(final String indent) {
+     *
+     * @return the {@link StreamsMetadata}s for the local instance in a {@link KafkaStreams} application
+     */
+    public synchronized StreamsMetadata getMyMetadata() {
```
"My" is an ambiguous term. Maybe "getLocalMetadata" or something?

vinothchandar replied:
yes `Local` is better and matches the lag API also

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -127,6 +137,7 @@ public String toString(final String indent) {
      * @return The {@link StreamsMetadata} for the storeName and key or {@link StreamsMetadata#NOT_AVAILABLE}
      * if streams is (re-)initializing
      */
+    @Deprecated
```
When adding deprecations, please also add the javadoc for them, including what version they were deprecated in and what should be used instead.

vinothchandar replied:
Agree. Will point out alternative.. I dont think we should we mix version information here, since we could backport something, this may be distributed differently etc. 
 KafkaStreams#close method for e.g does not do that.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -158,9 +169,69 @@ public String toString(final String indent) {
+        }
 
+        if (globalStores.contains(storeName)) {
+            // global stores are on every node. if we dont' have the host info
```
```suggestion
            // global stores are on every node. if we don't have the host info
```

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +298,89 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
+        final Map<String, List<String>> storeToSourceTopics = builder.stateStoreNameToSourceTopics();
+        Stream.concat(activePartitionHostMap.keySet().stream(), standbyPartitionHostMap.keySet().stream())
+            .distinct()
+            .forEach(key -> {
```
maybe `hostInfo` would be a clearer name.

vinothchandar replied:
agree. just used how it was before. will change

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +298,89 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
+        return new KeyQueryMetadata(activeHost, standbyHosts, partition);
     }
 
+    @Deprecated
```
clarification on the earlier comment: I wouldn't bother adding the deprecation javadoc for internal methods that don't already have a doc, unless you feel like it enhances clarity.

vinothchandar replied:
I think it may help streams devs and give them a note on writing future code calling this. I am fine removing if you feel strongly.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:
```diff
@@ -16,29 +16,31 @@
-import org.apache.kafka.streams.state.HostInfo;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import static org.apache.kafka.streams.processor.internals.assignment.StreamsAssignmentProtocolVersions.LATEST_SUPPORTED_VERSION;
```
Is it necessary to re-order these imports?

vinothchandar replied:
its not; unintentional; I think thats whats IntelliJ optimize imports did.. This is a common problem I have with another project as well. Anyway will revert to how it was

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/StreamsMetadata.java:
```diff
@@ -87,24 +117,29 @@ public boolean equals(final Object o) {
+            return false;
+        }
 
+        return standbyTopicPartitions.equals(that.standbyTopicPartitions);
```
These custom equals/hashCode methods make me nervous... Can we transition all the ones you've had to touch to use `Objects`?

vinothchandar replied:
Do you mean replace calling `a.equals(b)`, with `Objects.equals(a, b)`? It will give us null checking. so yes. Will change.. 
I don't follow "custom" adjective.. We have to compare each field right anyway right?

vinothchandar replied:
and FWIW its how it is on master as well https://github.com/apache/kafka/blob/39158bd944f7c64e32634cec8171515b93466ffa/streams/src/main/java/org/apache/kafka/streams/state/StreamsMetadata.java#L83 

Will make a sweep and change them

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java:
```diff
@@ -46,12 +48,18 @@ public StreamThreadStateStoreProvider(final StreamThread streamThread) {
+        if (!streamThread.isRunning()) {
             throw new InvalidStateStoreException("Cannot get state store " + storeName + " because the stream thread is " +
-                    streamThread.state() + ", not RUNNING");
+                streamThread.state() + ", not RUNNING");
```
```suggestion
                streamThread.state() + ", not RUNNING or REBALANCING");
```

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -659,15 +668,43 @@ public void shouldNotGetAllTasksWithStoreWhenNotRunning() {
+    }
+
+    @Test
+    public void shouldReturnEmptyLocalStorePartitionLags() {
```
because the topology is empty?

vinothchandar replied:
No tasks are running.. The streams threads themselves are mocked at this test level. I had a hard time trying to mock different pieces. This test is verifying that certain methods are being called to compute the lag. thats all... 

Real tests on lag values will happen on an integration test, that brings up a real topology. lmk if you have better suggestions/advice on testing this.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -321,15 +342,15 @@ private void verifyAllWindowedKeys(final List<KafkaStreams> streamsList,
-                        .metadataForKey(storeName, key, new StringSerializer());
-                    if (metadata == null || metadata.equals(StreamsMetadata.NOT_AVAILABLE)) {
+                    final KeyQueryMetadata metadata = streams
+                        .queryMetadataForKey(storeName, key, new StringSerializer());
```
should we also preserve the tests for the deprecated methods?

vinothchandar replied:
I pondered this early on. Seems some deprecated methods in KafkaStreams, just got their tests changed. I could call the other method as well `metadataForKey` and match it with the new API may be, to prevent just duplicating tests. wdyt

vinothchandar replied:
```
/Users/vchandar/Code/kafka/streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:352: warning: [deprecation] <K>metadataForKey(String,K,Serializer<K>) in KafkaStreams has been deprecated
```
ah yes. build fails when deprecated methods are called. Let me see if there is a way to supress this and have one call in the integration test that validates the new and old api against each other

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java:
```diff
@@ -122,28 +120,32 @@ public Object apply(final Object value) {
 
     @Test
-    public void shouldNotThrowNPEWhenOnChangeNotCalled() {
+    public void shouldNotThrowExceptionWhenOnChangeNotCalled() {
```
This is fine, but it seems like we should also make some assertions about what gets returned in this case.

vinothchandar replied:
This is just a small rename, while I was there, since the test name was checking for NPE, when it would fail if there was any exception..   Added an assert

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfoTest.java:
```diff
@@ -36,80 +36,110 @@
     };
-    private final Map<HostInfo, Set<TopicPartition>> globalAssignment = new HashMap<HostInfo, Set<TopicPartition>>() {
+    private final Map<HostInfo, Set<TopicPartition>> activeAssignment = new HashMap<HostInfo, Set<TopicPartition>>() {
+        {
```
We shouldn't use double-brace initialization. There's a `mkMap` utility function to use instead.

vinothchandar replied:
I tend to play along with how files are.  esp when I am new to this code base. This is how master is.. Will change these methods. I am using mkMap/mkEntry at other places.. its convenient.

vvcephei reviewed (commented):
Hey @vinothchandar , I've completed my high-level pass. I left comments where I had any concerns for you to follow up on.

One thing I didn't do is confirm that the public API changes match exactly what was proposed in the KIP. I assume you'll double-check that before calling for final reviews.

Can you also confirm that you've run the tests with coverage and all the new logic is being exercised by the tests? (There are a few unused methods that indicate testing gaps, but there could be more, hidden in branches). Also on this front, I'd just re-iterate that we should make sure that the lag-computation logic still works when the source-table optimization is enabled.

Thanks!

vvcephei commented:
@vinothchandar , when you're ready to call for final reviews, can you also kick off a few runs of the Streams system tests to be sure they still pass?

vinothchandar committed:
Cleanup code bases on CR comments

vinothchandar commented:
@vvcephei Thanks for the review. I made a note of the pending test coverage, KIP update and system test results in the checklist above. I intend to get past that before calling the final review. 

I have addressed most of your comments and pushed it up. I wanted to get a high level pass early on, mainly to get a confirmation that this code/implementation direction is on track and avoid surprises at the last minute.. e.g  all these maps we are accessing to get offset information, do they look correct? . Do you see anything here that needs to change a lot?

vinothchandar committed:
Bringing KafkaStreams#metadataForKey() back into QueryableStateIntegrationTest + few cleanups

vinothchandar committed:
Adding LagFetchIntegrationTest to test the lag values obtained during rebalancing/restoration

vinothchandar committed:
Small change to enable LagInfo#equals() for test coverage

vinothchandar commented:
@vvcephei @brary Just updated the status in the PR description.. I am currently planning to tackle the 1 failing system test and the source topic optimization case above and update the KIP with the final details. 

The optimization is per se not needed by ksql atm (not sure about navinder's use-case), I plan to spend some time on this to make it work. But, given my timelines/dealines, I would like to split it as a follow on if thats acceptable. (lag API will then throw exception when optimization is turned on for now)

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1116,125 @@ public void cleanUp() {
+     *
+     * @return map of store names to another map of partition to {@link LagInfo}s
+     */
+    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
```
@vinothchandar , I took a closer look at the computations in this method. Here's my recommendation:

```java
    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
        final long latestSentinel = -2L;
        final Map<String, Map<Integer, LagInfo>> localStorePartitionLags = new TreeMap<>();

        final Collection<TopicPartition> allPartitions = new LinkedList<>();
        final Map<TopicPartition, Long> changelogPositions = new HashMap<>();

        // Obtain the current positions, of all the active-restoring and standby tasks
        for (final StreamThread streamThread : threads) {
            for (final StandbyTask standbyTask : streamThread.allStandbyTasks()) {
                allPartitions.addAll(standbyTask.changelogPartitions());
                changelogPositions.putAll(standbyTask.restoredOffsets());
            }

            final Set<TaskId> restoringTaskIds = streamThread.restoringTaskIds();
            for (final StreamTask activeTask : streamThread.allStreamsTasks()) {
                final Collection<TopicPartition> taskChangelogPartitions = activeTask.changelogPartitions();
                allPartitions.addAll(taskChangelogPartitions);

                final boolean isRestoring = restoringTaskIds.contains(activeTask.id());
                final Map<TopicPartition, Long> restoredOffsets = activeTask.restoredOffsets();
                for (final TopicPartition topicPartition : taskChangelogPartitions) {
                    if (isRestoring && restoredOffsets.containsKey(topicPartition)) {
                        changelogPositions.put(topicPartition, restoredOffsets.get(topicPartition));
                    } else {
                        changelogPositions.put(topicPartition, latestSentinel);
                    }
                }
            }
        }

        log.debug("Current changelog positions: {}", changelogPositions);
        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets;
        try {
            allEndOffsets = adminClient.listOffsets(
                allPartitions.stream()
                             .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()))
            ).all().get();
        } catch (final RuntimeException | InterruptedException | ExecutionException e) {
            throw new StreamsException("Unable to obtain end offsets from kafka", e);
        }
        log.debug("Current end offsets :{}", allEndOffsets);
        for (final Map.Entry<TopicPartition, ListOffsetsResultInfo> entry : allEndOffsets.entrySet()) {
            // Avoiding an extra admin API lookup by computing lags for not-yet-started restorations
            // from zero instead of the real "earliest offset" for the changelog.
            // This will yield the correct relative order of lagginess for the tasks in the cluster,
            // but it is an over-estimate of how much work remains to restore the task from scratch.
            final long earliestOffset = 0L;
            final long changelogPosition = changelogPositions.getOrDefault(entry.getKey(), earliestOffset);
            final long latestOffset = entry.getValue().offset();
            final LagInfo lagInfo = new LagInfo(changelogPosition == latestSentinel ? latestOffset : changelogPosition, latestOffset);
            final String storeName = streamsMetadataState.getStoreForChangelogTopic(entry.getKey().topic());
            localStorePartitionLags.computeIfAbsent(storeName, ignored -> new TreeMap<>())
                                   .put(entry.getKey().partition(), lagInfo);
        }

        return Collections.unmodifiableMap(localStorePartitionLags);
    }
```

vinothchandar replied:
@vvcephei Trying to diff and understand what the changes are.. Will respond in a bit

vinothchandar replied:
Changes seem to be :
- using a `TreeMap`,  (I guess for determinisitic ordering?)
- Use plain for loops
- Changing of sentinel.. 

Don't know if I follow why the following is needed
```
            final long changelogPosition = changelogPositions.getOrDefault(entry.getKey(), earliestOffset);
```
To handle the case, where is a changelog partition for standby without any `restoredOffsets`? Could you give me an example? I think the standby does not go through the restore path, so we may not find stuff in the restoredOffsets map? May be we could just track the last offset seen in  `StandbyTask#update` and just use that value directly? (does not handle source topic optimization). Your concerns on the EOS scenario w checkpoints is valid.

vinothchandar replied:
We plan to use `ProcessStateManager#standbyRestoredOffsets` for this.. We agreed `restoredOffsets` wont cut it for standbys

brary commented:
> @vvcephei @brary Just updated the status in the PR description.. I am currently planning to tackle the 1 failing system test and the source topic optimization case above and update the KIP with the final details.
> 
> The optimization is per se not needed by ksql atm (not sure about navinder's use-case), I plan to spend some time on this to make it work. But, given my timelines/dealines, I would like to split it as a follow on if thats acceptable. (lag API will then throw exception when optimization is turned on for now)

Hi @vinothchandar, thanks and looks reasonable to me. Even for me this works fine as it is.

vvcephei commented:
Hey all, to try and decrease the scope of this PR, I pulled out the part I feel most comfortable with merging as https://github.com/apache/kafka/pull/7960 . Please take a look and make sure it still looks good to you (@vinothchandar and @brary ). I plan to kick off the system tests and merge that part of it after they pass.

vvcephei commented:
Hi all, I've factored out and merged two big parts of this PR:
1. the ability to query standbys (#7962)
2. the extensions to the cluster metadata API (#7960)

Once this PR is rebased, it should only include changes to computing and reporting the lags on local stores as proposed in KIP-535.

vinothchandar commented:
There is a separate PR for this already.. #7961 

if everyone agrees, we can close this and just focus on that

brary commented:
Since other PR's have been merged. We can close this one.
